### PR TITLE
Dynamic parameter for mountpaths

### PR DESCRIPTION
**Pull request by AnPucel**
This query parameter will be primarily for DevEx and code generation usage. The idea is to replace hard coded mount paths in the OpenApi spec with a dynamic parameter i.e. `{mountPath}`.

When the new call is made to `/sys/internal/specs/openapi?generic_paths=true` the resulting OpenApi spec will have 

`"paths": {
    "/auth/{mountPath}/login/": {
      "description": "",
      "parameters": [
        {
          "name": "mountPath",
          "description": "Path that the backend was mounted at",
          "in": "path",
          "schema": {
            "type": "string"
          },
          "required": true
        }
      ],`

The resulting generated code will allow users of the client library to pass through a mount path that would've otherwise been hardcoded as e.g. `approle`

### CODE REVIEW

File: sdk/framework/backend.go [hunk 1]
```diff
@@ -527,10 +528,12 @@ func (b *Backend) handleRootHelp(req *logical.Request) (*logical.Response, error
 	// the request, the type will be used as part of the request/response body
 	// names in the OAS document.
 	requestResponsePrefix := req.GetString("requestResponsePrefix")
+	dynamicPaths := fmt.Sprintf("%v", req.Data["dynamicPaths"])
```
[the remark this answers is not in the document]
Comment (averche) [reply sdk/framework/backend.go:531]: You may want to consider `req.GetString` wrapper instead, similar to the one above

File: vault/logical_system_paths.go [hunk 2]
```diff
@@ -900,6 +900,11 @@ func (b *SystemBackend) internalPaths() []*framework.Path {
 					Type:        framework.TypeString,
 					Description: "Context string appended to every operationId",
 				},
+				"dynamic": {
```
Comment (averche) [inline vault/logical_system_paths.go:903]: I think this might need updating as well.

File: vault/logical_system_test.go [hunk 3]
```diff
@@ -3408,6 +3408,59 @@ func TestSystemBackend_InternalUIMount(t *testing.T) {
...
+		}
+		tag := doc.Paths[path.path].Get.Tags[0]
+		if tag != path.tag {
+			t.Fatalf("path: %s; expected tag: %s, actual: %s", path.path, tag, path.tag)
+		}
+	}
+
+	// Simple sanity check of response size (which is much larger than most
+	// Vault responses), mainly to catch mass omission of expected path data.
+	minLen := 70000
```
Comment (averche) [inline vault/logical_system_test.go:3458]:
```suggestion
	const minLen = 70000
```

File: vault/logical_system.go [hunk 4]
```diff
@@ -4008,6 +4008,8 @@ func (b *SystemBackend) pathInternalOpenAPI(ctx context.Context, req *logical.Re
 	// be received from plugin backends.
 	doc := framework.NewOASDocument()
 
+	dynamic := d.Get("dynamic").(bool)
```
[the remark this answers is not in the document]
Comment (AnPucel) [reply vault/logical_system.go:4011]: This is the parameter you append on to the request. So, I'm open to different names here. We can do `dynamicPaths` instead.

File: sdk/framework/backend.go [hunk 5]
```diff
@@ -527,10 +528,12 @@ func (b *Backend) handleRootHelp(req *logical.Request) (*logical.Response, error
...
+	dynamic, _ := strconv.ParseBool(dynamicPaths)
 
 	// Build OpenAPI response for the entire backend
 	doc := NewOASDocument()
-	if err := documentPaths(b, requestResponsePrefix, doc); err != nil {
+	if err := documentPaths(b, requestResponsePrefix, dynamic, doc); err != nil {
```
[the remark this answers is not in the document]
Comment (AnPucel) [reply sdk/framework/backend.go:536]: I feel like `custom` is a little confusing. B/c you mount engines at a `custom` path, so it almost sounds like the opposite to me. Like it would use the hardcoded ones in this case.

we could do `dynamicMountPaths` and update everything accordingly?

File: sdk/framework/openapi.go [hunk 6]
```diff
@@ -213,18 +213,20 @@ var (
...
-		if err := documentPath(p, backend.SpecialPaths(), requestResponsePrefix, backend.BackendType, doc); err != nil {
+		if err := documentPath(p, backend.SpecialPaths(), requestResponsePrefix, dynamicPaths, backend.BackendType, doc); err != nil {
 			return err
 		}
 	}
 
 	return nil
 }
 
+// Set requestResponsePrefix = {mountPath}
```
Comment (AnPucel) [inline sdk/framework/openapi.go:226]: remove comments

File: sdk/framework/backend.go [hunk 7]
```diff
@@ -527,10 +528,12 @@ func (b *Backend) handleRootHelp(req *logical.Request) (*logical.Response, error
 	// the request, the type will be used as part of the request/response body
 	// names in the OAS document.
 	requestResponsePrefix := req.GetString("requestResponsePrefix")
+	genericPaths := fmt.Sprintf("%v", req.Data["genericPaths"])
+	generic, _ := strconv.ParseBool(genericPaths)
```
Comment (averche) [inline sdk/framework/backend.go:532]: I think it might be still worth specifying that it's "mount" paths (since we are dealing with many different paths in Vault).

Comment (AnPucel) [reply sdk/framework/backend.go:532]: sounds good! If you think that's not too long. Do you think the query parameter should have `mount` as well? It seems a little long, but I don't have a strong opinion either way

File: sdk/framework/backend.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (AnPucel) [reply sdk/framework/backend.go]: I believe that didn't work in this case. But let me try again, maybe I was using it incorrectly

File: sdk/framework/backend.go (hunk 7 above)
Comment (raskchanky) [reply sdk/framework/backend.go:532]: Why are we ignoring the error from `strconv.ParseBool()`?

File: vault/logical_system.go [hunk 8]
```diff
@@ -4079,7 +4081,11 @@ func (b *SystemBackend) pathInternalOpenAPI(ctx context.Context, req *logical.Re
 					}
 				}
 
-				doc.Paths["/"+mountPrefix+mount+path] = obj
+				if genericPaths && mount != "sys/" && mount != "identity/" {
+					doc.Paths["/"+mountPrefix+"{mountPath}/"+path] = obj
```
Comment (raskchanky) [inline vault/logical_system.go:4085]: I had to stare at this for a few minutes 😆 so just to make sure I'm following - `{mountPath}` is a string literal here, yes? Not some kind of variable interpolation I'm not familiar with?

Comment (AnPucel) [reply vault/logical_system.go:4085]: Haha. Yes! it is just a string literal here. That's how we want it to look in the spec,  but I can add a comment for clarification.

Comment (averche) [reply vault/logical_system.go:4085]: Using `fmt.Sprintf` to construct the string might make it more readable

File: sdk/framework/backend.go (hunk 7 above)
Comment (AnPucel) [reply sdk/framework/backend.go]: I just fixed the parsing bool to be more straightfoward

File: sdk/framework/backend.go [hunk 9]
```diff
@@ -527,10 +527,11 @@ func (b *Backend) handleRootHelp(req *logical.Request) (*logical.Response, error
 	// the request, the type will be used as part of the request/response body
 	// names in the OAS document.
 	requestResponsePrefix := req.GetString("requestResponsePrefix")
+	genericPaths := req.Get("genericPaths").(bool)
```
Comment (raskchanky) [inline sdk/framework/backend.go:530]: What happens if `genericPaths` isn't passed as a parameter? Does this panic on the type assertion?

File: vault/logical_system.go [hunk 10]
```diff
@@ -4008,6 +4008,8 @@ func (b *SystemBackend) pathInternalOpenAPI(ctx context.Context, req *logical.Re
 	// be received from plugin backends.
 	doc := framework.NewOASDocument()
 
+	genericPaths := d.Get("generic_paths").(bool)
```
Comment (raskchanky) [inline vault/logical_system.go:4011]: What happens if `generic_paths` isn't passed as a parameter? Does this panic on the type assertion?

File: sdk/framework/backend.go (hunk 9 above)
Comment (AnPucel) [reply sdk/framework/backend.go:530]: oops. yep. this one does. Let me fix that

File: sdk/framework/backend.go [hunk 11]
```diff
@@ -528,9 +528,20 @@ func (b *Backend) handleRootHelp(req *logical.Request) (*logical.Response, error
 	// names in the OAS document.
 	requestResponsePrefix := req.GetString("requestResponsePrefix")
 
+	// Generic paths will primarily be used for code generation purposes.
+	// This will result in a dynamic mount path being placed instead of
+	// a hardcoded default path. For example /auth/approle/login would be replaced
```
Comment (averche) [inline sdk/framework/backend.go:533]:
```suggestion
	// Generic mount paths will primarily be used for code generation purposes.
	// This will result in a dynamic mount paths being placed instead of
	// a hardcoded default paths. For example /auth/approle/login would be replaced
```

File: sdk/framework/backend.go [hunk 12]
```diff
@@ -528,9 +528,20 @@ func (b *Backend) handleRootHelp(req *logical.Request) (*logical.Response, error
...
+	// with /auth/{mountPath}/login. This will be replaced for all secrets
+	// engines and auth methods that are enabled.
+	genericMountPaths, ok := req.Get("genericMountPaths").(bool)
```
Comment (averche) [inline sdk/framework/backend.go:536]: Would this typecasting work if the API endpoint is called through curl (`sys/internal/specs/openapi`)?generic_mount_paths=true`. I'm assuming "true" here is a string and it would be typecast into a bool, which would result in `false`.

File: sdk/framework/backend.go [hunk 13]
```diff
@@ -528,9 +528,20 @@ func (b *Backend) handleRootHelp(req *logical.Request) (*logical.Response, error
...
+
+	if !ok {
+		genericMountPaths = false
+	}
```
Comment (averche) [inline sdk/framework/backend.go:540]: If the typecasting works, I think this part is redundant, since false is the default value
```suggestion
```

File: website/content/api-docs/system/internal-specs-openapi.mdx [hunk 14]
```diff
@@ -31,6 +31,10 @@ This endpoint returns a single OpenAPI document describing all paths visible to
 | :----- | :---------------------------- |
 | `GET`  | `/sys/internal/specs/openapi` |
 
+### Parameters
+
+- `generic_mount_paths` `(bool: false)` – Used to specify whether to use a generic mount path. If set, the mount path will be replaced with a dynamic parameter: `{mountPath}`
```
Comment (averche) [inline website/content/api-docs/system/internal-specs-openapi.mdx:36]:
```suggestion
- `generic_mount_paths` `(bool: false)` – Used to specify whether to use a generic mount paths. If set, the mount paths will be replaced with a dynamic parameter: `{mountPath}`
```

It may be good to include an example here as well

File: sdk/framework/backend.go (hunk 13 above)
Comment (VinnyHC) [reply sdk/framework/backend.go:540]: @averche If you omit an `ok` check wouldn't this panic if the casting fails?

Comment (AnPucel) [reply sdk/framework/backend.go:540]: Yes, so this will fail without it when calling through the help endpoint, i.e. if it doesn't passthrough the `pathInternalOpenApi` call in logical_system.go. E.g. you're calling `http://127.0.0.1:8200/v1/auth/approle\?help\=1`

That gets handled differently b/c it's getting the mount path from the FieldData which is treating it like a bool, and there is a helper that will return the default value as specified in the FieldData for this. Thus why we need a check here and not in pathInternalOpenApi get call.

File: sdk/framework/backend.go (hunk 12 above)
Comment (averche) [reply sdk/framework/backend.go:536]: Talked offline and this should not be a point since the underlying type of `req.Get("genericMountPaths")` is `bool`

File: vault/logical_system.go [hunk 15]
```diff
@@ -4008,6 +4008,8 @@ func (b *SystemBackend) pathInternalOpenAPI(ctx context.Context, req *logical.Re
 	// be received from plugin backends.
 	doc := framework.NewOASDocument()
 
+	genericMountPaths := d.Get("generic_mount_paths").(bool)
```
Comment (averche) [inline vault/logical_system.go:4011]:
```suggestion
	genericMountPaths, _ := d.Get("generic_mount_paths").(bool)
```

One more to avoid panic :)

Comment (AnPucel) [reply vault/logical_system.go:4011]: I think this particular `Get` guards against this if the field doesn't exist and uses the type default instead. But I'm happy to add it for security sake. haha

Comment (averche) [reply vault/logical_system.go]: I think what we want to guard againt is the type being a non-bool (e.g. a string) which could cause a panic.

Comment (AnPucel) [reply vault/logical_system.go]: okay. I'm confused. If it is a string, doesn't acknowledging the `Ok` make it default to false? thus avoiding the panic? Maybe I misunderstood the discussion.

Also, b/c the field data sets this `generic_mount_paths` as a boolean, it will come back as a boolean w/ this particular `Get` as it decodes the values into the proper types. There's no way for it to come through as a string.

Comment (AnPucel) [reply vault/logical_system.go]: oooh. wait. sorry. I misunderstood. I did add the suggested fix

File: website/content/api-docs/system/internal-specs-openapi.mdx (hunk 14 above)
Comment (AnPucel) [reply website/content/api-docs/system/internal-specs-openapi.mdx]: done

**Review by averche — APPROVED** [review]
LGTM
